### PR TITLE
Custom cna namespace annotation json (RFC 65)

### PR DESCRIPTION
**Pull request by BasLee**
Add custom namespace columns to CNA events

image

## Changes
- Add `ANNOTATION_JSON` column to `sample_cna_event table`
- Add [CustomNamespaceParser.java](https://github.com/BasLee/cbioportal/compare/custom-cna-namespace...BasLee:cbioportal:custom-cna-namespace-annotation-json?expand=1#diff-8133c497150e14bfb30c8c96682d8ef8cdb2f37e32784592706224c0a8179283) that imports custom columns of both mutations and CNA's
- Expose the custom columns as `namespaceColumns` in the `DiscreteCopyNumberController` (see `DiscreteCopyNumberDataMixin`)

## Tests
- Cna importer imports custom namespace columns, see: `TestImportCnaDiscreteLongData`
- API exposes custom namespace columns, see: `DiscreteCopyNumberServiceImplTest` and `DiscreteCopyNumberControllerTest`

# Checks
- [ ] When importing es0, custom namespace columns should be added to the response of [./api/molecular-profiles/study_es_0_gistic/discrete-copy-number](http://0.0.0.0:8081/api/molecular-profiles/study_es_0_gistic/discrete-copy-number?sampleListId=study_es_0_cna&discreteCopyNumberEventType=HOMDEL_AND_AMP) :
 ```
response.filter(e => e.entrezGeneId === 116983 && e.sampleId === 'TCGA-A2-A04U-01').map(e => e.namespaceColumns)
{
    "MyNamespace": {
        "column1": "value1b",
        "column2": "value2b"
    },
    "MyNamespace2": {
        "blarp": "bluirp2"
    }
}
```

### CODE REVIEW

File: core/src/main/java/org/mskcc/cbio/maf/MafUtil.java [hunk 1]
```diff
@@ -215,7 +210,7 @@ public class MafUtil {
 
     // mapping for all column names (both standard and custom columns)
     private HashMap<String, Integer> columnIndexMap;
-    private Map<String, Map<String, Integer>> namespaceIndexMap;
+    private final CustomNamespaceParser customNamespaceParser;
```
Comment (pvannierop) [inline core/src/main/java/org/mskcc/cbio/maf/MafUtil.java]:
```suggestion
    private final NamespaceColumnParser namespaceColumnParser;
```

File: core/src/main/java/org/mskcc/cbio/maf/MafUtil.java [hunk 2]
```diff
@@ -485,30 +464,15 @@ public MafRecord parseRecord(String line) {
...
+        record.setNamespacesMap(
+            this.customNamespaceParser.parseCustomNamespaces(parts)
+        );
 
         return record;
     }
 
     public Map<String, Map<String, Integer>> getNamespaceIndexMap() {
-        return namespaceIndexMap;
+        return this.customNamespaceParser.getNamespaceIndexMap();
```
Comment (pvannierop) [inline core/src/main/java/org/mskcc/cbio/maf/MafUtil.java]:
```suggestion
        return this.customNamespaceParser.getNamespaceColumnIndexMap();
```

File: core/src/main/scripts/importer/cbioportalImporter.py [hunk 3]
```diff
@@ -100,9 +100,9 @@ def remove_study_id(jvm_args, study_id):
     args.append(study_id)
     args.append("--noprogress") # don't report memory usage and % progress
     run_java(*args)
-
-
-def import_study_data(jvm_args, meta_filename, data_filename, update_generic_assay_entity, meta_file_dictionary = None):
+    
+
```
Comment (pvannierop) [inline core/src/main/scripts/importer/cbioportalImporter.py]: Remove whitespace

File: core/src/main/scripts/importer/cbioportalImporter.py [hunk 4]
```diff
@@ -327,6 +328,9 @@ def process_directory(jvm_args, study_directory, update_generic_assay_entity = N
         elif meta_file_type == MetaFileTypes.STRUCTURAL_VARIANT:
             structural_variant_filepair = (
                 (meta_filename, os.path.join(study_directory, meta_dictionary['data_filename'])))
+        elif meta_file_type == MetaFileTypes.CNA_DISCRETE:
+            cna_long_filepair = (
+                (meta_filename, os.path.join(study_directory, meta_dictionary['data_filename'])))
```
Comment (pvannierop) [inline core/src/main/scripts/importer/cbioportalImporter.py]: Should there not be a check that the data provided is in LONG format. E.g.:
```
MetaFileTypes.CNA_DISCRETE_LONG:
```

File: core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java [hunk 5]
```diff
@@ -91,8 +92,8 @@ public void testImportCnaDiscreteLongDataAddsSamples() throws Exception {
             geneticProfile.getGeneticProfileId(),
             genePanel,
             DaoGeneOptimized.getInstance(),
-            DaoGeneticAlteration.getInstance()
-        ).importData();
+            DaoGeneticAlteration.getInstance(),
+            noNamespaces).importData();
```
Comment (pvannierop) [inline core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java:96]: Nice that you add a human readable shorthand! Well done!

File: core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java [hunk 6]
```diff
@@ -284,6 +287,98 @@ public void testImportCnaDiscreteLongDataAddsPdAnnotations() throws Exception {
...
 
+    /**
+     * Test the import of cna data file in long format with:
+     * - two custom namespaces and one unknown namespace
+     */
+    @Test
+    public void testImportCnaDiscreteLongDataAddsCustomNamespaceColumns() throws Exception {
+        ObjectMapper mapper = new ObjectMapper();
+        File file = new File("src/test/resources/data_cna_discrete_import_test_with_namespaces.txt");
+        Set<String> namespaces = newHashSet("MyNamespace", "MyNamespace2");
```
Comment (pvannierop) [inline core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java]:
```suggestion
        Set<String> namespacesToImport = newHashSet("MyNamespace", "MyNamespace2");
```

File: core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java [hunk 7]
```diff
@@ -284,6 +287,98 @@ public void testImportCnaDiscreteLongDataAddsPdAnnotations() throws Exception {
         assertEquals("Class 2,Class 1,NA", allDriverTiersFilters);
     }
 
+    /**
+     * Test the import of cna data file in long format with:
+     * - two custom namespaces and one unknown namespace
+     */
```
Comment (pvannierop) [inline core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java]: It is not clear to me from use of variable names or comments what is tested in this test. Especially, the part where the comment states `Only one namespace column provided`. The Javadoc at the top suggests that this is the case `unknown namespace`? Can you make this more clear? Two separate tests perhaps?

File: core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java [hunk 8]
```diff
@@ -311,6 +406,32 @@ private void assertSampleExistsInGeneticProfile(String sampleId) throws DaoExcep
         ));
     }
 
+    class AnnoJson {
```
Comment (pvannierop) [inline core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java]:
```suggestion
    class NamespaceAnnotationJson {
```

File: service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java [hunk 9]
```diff
@@ -101,10 +101,10 @@ public List<DiscreteCopyNumberData> getDiscreteCopyNumbersInMultipleMolecularPro
...
-            sampleIds,
-            entrezGeneIds,
-            projection)
+                molecularProfileIds,
+                sampleIds,
+                entrezGeneIds,
+                projection)
             .stream()
             .filter(g -> isValidAlteration(alterationTypes, g))
             .map(this::convert)
```
Comment (pvannierop) [inline service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java:115]: Please correct indentation (make consistent).

File: service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java [hunk 10]
```diff
@@ -120,14 +120,14 @@ public List<DiscreteCopyNumberData> getDiscreteCopyNumbersInMultipleMolecularPro
...
+
         if (isHomdelOrAmpOnlyCna(cnas)) {
             return discreteCopyNumberRepository.getDiscreteCopyNumbersInMultipleMolecularProfilesByGeneQueries(molecularProfileIds,
                 sampleIds, geneQueries, projection);
         }
-        
+
         return molecularDataService.getMolecularDataInMultipleMolecularProfilesByGeneQueries(molecularProfileIds, sampleIds,
-            geneQueries, projection).stream()
+                geneQueries, projection).stream()
```
Comment (pvannierop) [inline service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java]: Please correct indentation (make consistent).

File: web/src/main/java/org/cbioportal/web/config/CustomObjectMapper.java [hunk 11]
```diff
@@ -39,68 +39,16 @@
...
-import org.cbioportal.model.GisticToGene;
-import org.cbioportal.model.MolecularProfile;
-import org.cbioportal.model.MutSig;
-import org.cbioportal.model.Mutation;
-import org.cbioportal.model.MutationSpectrum;
-import org.cbioportal.model.Patient;
-import org.cbioportal.model.ResourceDefinition;
-import org.cbioportal.model.Sample;
-import org.cbioportal.model.SampleList;
+import org.cbioportal.model.*;
```
Comment (pvannierop) [inline web/src/main/java/org/cbioportal/web/config/CustomObjectMapper.java]: Better to restore the old import format.

**Review by pvannierop — CHANGES_REQUESTED** [review]
Great work. My compliments!

File: web/src/main/java/org/cbioportal/web/config/CustomObjectMapper.java (hunk 11 above)
Comment (BasLee) [reply web/src/main/java/org/cbioportal/web/config/CustomObjectMapper.java]: Fixed in [e741f34](https://github.com/BasLee/cbioportal/pull/1/commits/e741f34f8a023df66b841a406d323e650cdf2308)

File: service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java (hunk 10 above)
Comment (BasLee) [reply service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java]: Fixed in [9f7f6a5](https://github.com/BasLee/cbioportal/pull/1/commits/9f7f6a57df2238791075041f36851b087cdd415f)

File: service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java (hunk 9 above)
Comment (BasLee) [reply service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java:115]: Fixed in [9f7f6a5](https://github.com/BasLee/cbioportal/pull/1/commits/9f7f6a57df2238791075041f36851b087cdd415f)

File: core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java (hunk 8 above)
Comment (BasLee) [reply core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java]: Fixed in [9f7f6a5](https://github.com/BasLee/cbioportal/pull/1/commits/9f7f6a57df2238791075041f36851b087cdd415f)

File: core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java (hunk 7 above)
Comment (BasLee) [reply core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java]: Fixed in [9f7f6a5](https://github.com/BasLee/cbioportal/pull/1/commits/9f7f6a57df2238791075041f36851b087cdd415f)

File: core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java (hunk 6 above)
Comment (BasLee) [reply core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportCnaDiscreteLongData.java]: Fixed in [9f7f6a5](https://github.com/BasLee/cbioportal/pull/1/commits/9f7f6a57df2238791075041f36851b087cdd415f)

File: core/src/main/scripts/importer/cbioportalImporter.py (hunk 4 above)
Comment (BasLee) [reply core/src/main/scripts/importer/cbioportalImporter.py]: Fixed in [9f7f6a5](https://github.com/BasLee/cbioportal/pull/1/commits/9f7f6a57df2238791075041f36851b087cdd415f)

File: core/src/main/scripts/importer/cbioportalImporter.py (hunk 3 above)
Comment (BasLee) [reply core/src/main/scripts/importer/cbioportalImporter.py]: Fixed in [9f7f6a5](https://github.com/BasLee/cbioportal/pull/1/commits/9f7f6a57df2238791075041f36851b087cdd415f)

File: core/src/main/java/org/mskcc/cbio/maf/MafUtil.java (hunk 2 above)
Comment (BasLee) [reply core/src/main/java/org/mskcc/cbio/maf/MafUtil.java]: Fixed in [9f7f6a5](https://github.com/BasLee/cbioportal/pull/1/commits/9f7f6a57df2238791075041f36851b087cdd415f)

File: core/src/main/java/org/mskcc/cbio/maf/MafUtil.java (hunk 1 above)
Comment (BasLee) [reply core/src/main/java/org/mskcc/cbio/maf/MafUtil.java]: Fixed in [9f7f6a5](https://github.com/BasLee/cbioportal/pull/1/commits/9f7f6a57df2238791075041f36851b087cdd415f)

**Pull request closed without merging**